File: src/engine/Entity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;

namespace SimpleEngine
{
    public class Entity
    {
        public List<Renderable> visual = new List<Renderable>();
        public readonly string name;

        private Entity(string name)
        {
            this.name = name;
        }

        public override string ToString()
        {
            return name;
        }

        internal static Entity Create(MediaLoader loader, string name, System.Xml.XmlElement root, vec3 pos, quat rot)
        {
            Entity ent = new Entity(name);

            XmlElement visual = root["visual"];

            foreach(XmlElement meshel in Xml.ElementsNamed(visual, "mesh"))
            {
                string meshpath = Xml.GetAttributeString(meshel, "file");
                MeshInstance mesh = new MeshInstance(loader.fetch<Mesh>(meshpath));
                mesh.pos = pos;
                mesh.rot = rot;
                ent.add(mesh);
            }

            return ent;
        }

        private void add(Renderable r)
        {
            visual.Add(r);
        }
    }
}

File: src/engine/entity.cc
#include "engine/entity.h"

#include <memory>
#include <string>
#include <vector>

#include "engine/medialoader.h"
#include "engine/mesh.h"
#include "engine/meshinstance.h"
#include "engine/quat.h"
#include "engine/vec3.h"

namespace simple
{
    Entity::Entity(const std::string& n)
        : name(n)
    {
    }

    std::string Entity::ToString() const
    {
        return name;
    }

    std::shared_ptr<Entity> Entity::Create(MediaLoader* loader, const std::string& name, std::shared_ptr<xml::Element> root, const vec3& pos, const quat& rot)
    {
        auto ent = std::make_shared<Entity>(name);

        auto visual = root->GetChild("visual");

        for (auto meshel : xml::ElementsNamed(visual, "mesh"))
        {
            std::string meshpath = xml::GetAttributeString(meshel, "file");
            auto mesh = std::make_shared<MeshInstance>(loader->GetOrLoad<Mesh>(meshpath));
            mesh->position = pos;
            mesh->rotation = rot;
            ent->AddRenderable(mesh);
        }

        return ent;
    }

    void Entity::AddRenderable(std::shared_ptr<Renderable> r)
    {
        renderables.emplace_back(r);
    }
}
